internal/server/app.go

```
@@ -0,0 +1,242 @@
+				fmt.Sprintf(
+					"Unable to perform requested action on metrics type %s", req.MType,
+				),
+			),
+		)
+	}
+
+	value, err := app.store.Extract(ctx, req)
+	if err != nil {
+		return schema.NewEmptyMetrics(), convertError(err)
+	}
+
+	if app.key != "" {
+		if err = value.Sign(app.key); err != nil {
+			return schema.NewEmptyMetrics(), convertError(err)
+		}
+	}
+	return value, nil
+}
+
+func (app *App) updateValue(
+	ctx context.Context,
+	req schema.Metrics,
+) (
+	schema.Metrics,
+	*applicationError,
+) {
+	var err error
+
+	if (req.MType == schema.MetricsTypeCounter && req.Delta == nil) || (req.MType == schema.MetricsTypeGauge && req.Value == nil) {
+		return schema.NewEmptyMetrics(), convertError(
+			ValidationError(
+				"missing value",
+			),
+		)
+	}
+
+	if app.key != "" {
+		signed, e := req.IsSignedWithKey(app.key)
+		if e != nil {
+			return schema.NewEmptyMetrics(), convertError(e)
+		}
+		if !signed {
+			return schema.NewEmptyMetrics(), convertError(
+				ValidationError(
+					"signature mismatch",
+				),
+			)
+		}
+	}
+
+	switch req.MType {
+	case schema.MetricsTypeCounter:
+		err = app.store.Increment(ctx, req, *req.Delta)
+		switch err.(type) {
+		case *storage.NotFound:
+			err = app.store.Put(ctx, req)
+		}
+	case schema.MetricsTypeGauge:
+		err = app.store.Put(ctx, req)
+	default:
+		return schema.NewEmptyMetrics(), convertError(InvalidTypeError(
+			fmt.Sprintf(
+				"Unable to perform requested action on metrics type %s", req.MType,
+			),
+		))
+	}
+
+	if err != nil {
+		return schema.NewEmptyMetrics(), convertError(err)
+	}
+
+	if app.sync {
+		// by the logic of it, we should not depend on the request's context
+		// while syncing our state with dumper, so we use background context
+		go app.safeDump(context.Background())
+	}
+
+	return app.retrieveValue(ctx, req)
+}
+
+func (app *App) listValues(ctx context.Context) (
+	[]schema.Metrics,
+	*applicationError,
+) {
+	values, err := app.store.List(ctx)
+	if app.key != "" {
```

А здесь бы подошел fail fast / early exit